### ISSUE TITLE
Module version query failed

### ISSUE THREAD

**Issue by profmocs** [intel/thunderbolt-software-user-space#18]
I successfully built the thunderboltd. However, when I attempt to run it I get the following error.

`Error: Error: Exception: Module version query failed      Thread: 140085838165824, func main line 94 (/root/thunderbolt-software-user-space/daemon/ThunderboltService/Linux/thunderbolt.cpp)
`

I am using CentOS 7.

**Comment by profmocs** [comment]
This error occurs because it is looking for the thunderbolt_icm module. Where do I get that? The thunderbolt kernel tree builds a module called thunderbolt.

**Comment by ybernat** [comment]
If you want the module for the user-space components from master or fwupdate branches, please refer to https://github.com/01org/thunderbolt-software-kernel-tree/tree/networking

What TBT feature you are looking for? It's possible that you don't need either of them.

**Comment by michael-jamet** [comment]
@profmocs is your issue resolved?

**Comment by profmocs** [comment]
Sorry for the late response, I was only holiday. @ybernat, I am looking to do networking using TBT. @michael-jamet , my issue hasn't be resolved.  The Thunderbolt module isn't being loaded. I cannot determine why it isn't being loaded.

**Comment by ybernat** [comment]
@profmocs So yes, for TBT networking this is the correct solution for now.

Some more points to check:
How did you compile the kernel module? Have you compiled all the kernel or you compiled only the module?
Have you insmod/modprobe thunderbolt_icm?
Is anything connected to TBT port? (If not, the TBT controller is powered down.)

**Comment by profmocs** [comment]
1) I compiled the entire kernel
2) When I modprobe thunderbolt_icm I get "modprobe: FATAL: Module thunderbolt_icm not found" However, when I do "modprobe -v thunderbolt" I see the call to insmod, i.e. 
"insmod /lib/modules/4.11.0+/kernel/drivers/thunderbolt/thunderbolt.ko".
3) I have a hard drive connected.

**Comment by ybernat** [comment]
Have you selected `thunderbolt networking` in the config (e.g. with `make menuconfig`) before compiling the kernel?

**Comment by profmocs** [comment]
Yes, I did. I think that should be apparent from the presence of thunderbolt.ko in /lib/modules/4.11.0+. I will rebuild the kernel using the networking branch this time and see what happens.

**Comment by ybernat** [comment]
With these patches, `thunderbolt.ko` is still only for Apple systems (for device connection). `thunderbolt networking` is a separate setting that adds `thunderbolt_icm` module for TBT networking on PCs.

**Comment by profmocs** [comment]
OK. I figured out I didn't have the select the Thunderbolt Networking module. I selected and built the  kernel and the module loads with modprobe. However, I get the following error when running the daemon. `/usr/lib/thunderbolt/thunderboltd: symbol lookup error: /usr/lib/thunderbolt/thunderboltd: undefined symbol: _ZN4DBus13BusDispatcherC1Ev`.

**Comment by ybernat** [comment]
Googling this symbol, it looks like it comes from libdbus-c++.
What version of it is installed on your setup?
What is the output of `ldd /usr/lib/thunderbolt/thunderboltd`? Everything is found?

**Comment by profmocs** [comment]
I am guessing that the driver is built using a different version of libdbus-c++ than what I have installed. I have dbus 1.6.12 and dbus-c++ 0.5 on my box. According to the dbus web site, 1.6.12 is a legacy version. So I will upgrade to the latest dbus and dbus-c++ and see what happens.

**Comment by ybernat** [comment]
libdbus-c++ 0.5 should work fine.
What about ldd output?

**Comment by profmocs** [comment]
Here's the output of ldd.

linux-vdso.so.1 =>  (0x00007ffe445bd000)

libdbus-c++-1.so.0 => /lib64/libdbus-c++-1.so.0 (0x00007f7fbf2e3000)

libnl-genl-3.so.200 => /lib64/libnl-genl-3.so.200 (0x00007f7fbf0dd000)

libnl-3.so.200 => /lib64/libnl-3.so.200 (0x00007f7fbeebb000)

libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f7fbebb2000)

libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f7fbe996000)

libm.so.6 => /lib64/libm.so.6 (0x00007f7fbe693000)

libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f7fbe47d000)

libc.so.6 => /lib64/libc.so.6 (0x00007f7fbe0bc000)

libdbus-1.so.3 => /lib64/libdbus-1.so.3 (0x00007f7fbde73000)

libglib-2.0.so.0 => /lib64/libglib-2.0.so.0 (0x00007f7fbdb3c000)

libexpat.so.1 => /lib64/libexpat.so.1 (0x00007f7fbd912000)

/lib64/ld-linux-x86-64.so.2 (0x0000562b8605b000)

librt.so.1 => /lib64/librt.so.1 (0x00007f7fbd709000)

**Comment by profmocs** [comment]
I think I resolved the problem. I rebuilt thunderboltd and I started it and
get the following when I do

status thunderbolt.service. I also have the other end of the USB-C capable
in a Mac.

● thunderbolt.service - Thunderbolt Manager

   Loaded: loaded (/usr/lib/systemd/system/thunderbolt.service; static;
vendor preset: disabled)

   Active: active (running) since Tue 2017-08-08 22:34:31 EDT; 2min 13s ago

 Main PID: 3175 (thunderboltd)

   CGroup: /system.slice/thunderbolt.service

           └─3175 /usr/lib/thunderbolt/thunderboltd


Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]: 75 15 00 06 08 00 07
00 0c 00 00 00 0c 00 09 00 u...............

Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]: 01 00 00 08 5b 7b 38
b2                         ....[{8.

Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]:
---------------------------  END NETLINK MESSAGE   ----------------...-----

Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]: Debug: validate
netlink message: PASS      Thread: 140205992920832,....cpp)

Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]: Debug:
LinuxControllerCommandSender::OnEvent entry, PDF: 12      Th....cpp)

Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]: Debug: Message from
FW: PDF_FW_TO_SW_RESPONSE, Controller: 6001575 ....cpp)

Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]: Info: FW response:
INTER_DOMAIN_PACKET_SENT_RESPONSE, Controller: 0....cpp)

Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]: Info: Netlink
message received      Thread: 140205992920832, func O....cpp)

Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]: Debug:
LinuxControllerCommandSender::OnEvent exit, PDF: 12      Thr....cpp)

Aug 08 22:34:31 nuc1.praise-galore thunderboltd[3175]: Info:
GenlWrapper::cb_valid exit, message arrived      Thread: 1402....cpp)

Hint: Some lines were ellipsized, use -l to show in full.
